lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -1301,6 +1308,164 @@ ConventionAttr FMemModuleOp::getConventionAttr() {
+}
+
+ParseResult ClassOp::parse(OpAsmParser &parser, OperationState &result) {
+  auto *context = result.getContext();
+  auto &builder = parser.getBuilder();
+
+  // Parse the visibility attribute.
+  (void)mlir::impl::parseOptionalVisibilityKeyword(parser, result.attributes);
+
+  // Parse the name as a symbol.
+  StringAttr nameAttr;
+  if (parser.parseSymbolName(nameAttr, ::mlir::SymbolTable::getSymbolAttrName(),
+                             result.attributes))
+    return failure();
+
+  // Parse the module ports.
+  SmallVector<OpAsmParser::Argument> entryArgs;
+  SmallVector<Direction, 4> portDirections;
+  SmallVector<Attribute, 4> portNames;
+  SmallVector<Attribute, 4> portTypes;
+  SmallVector<Attribute, 4> portAnnotations;
+  SmallVector<Attribute, 4> portSyms;
+  SmallVector<Attribute, 4> portLocs;
+  if (parseModulePorts(parser, /*hasSSAIdentifiers=*/true, entryArgs,
+                       portDirections, portNames, portTypes, portAnnotations,
+                       portSyms, portLocs))
+    return failure();
+
+  // Ports on ClassOp cannot have annotations
+  for (auto annos : portAnnotations)
+    if (!cast<ArrayAttr>(annos).empty())
+      return failure();
+
+  // If attributes are present, parse them.
+  if (parser.parseOptionalAttrDictWithKeyword(result.attributes))
+    return failure();
+
+  assert(portNames.size() == portTypes.size());
+
+  // Record the argument and result types as an attribute.  This is necessary
+  // for external modules.
+
+  // Add port directions.
+  if (!result.attributes.get("portDirections"))
+    result.addAttribute("portDirections",
+                        direction::packAttribute(context, portDirections));
+
+  // Add port names.
+  if (!result.attributes.get("portNames"))
+    result.addAttribute("portNames", builder.getArrayAttr(portNames));
+
+  // Add the port types.
+  if (!result.attributes.get("portTypes"))
+    result.addAttribute("portTypes", builder.getArrayAttr(portTypes));
+
+  // Add the port symbols.
+  if (!result.attributes.get("portSyms")) {
+    FModuleLike::fixupPortSymsArray(portSyms, builder.getContext());
+    result.addAttribute("portSyms", builder.getArrayAttr(portSyms));
+  }
+
+  // Add port locations.
+  if (!result.attributes.get("portLocations"))
+    result.addAttribute("portLocations", ArrayAttr::get(context, portLocs));
+
+  // Notably missing compared to other FModuleLike, we do not track port
+  // annotations, nor port symbols, on classes.
+
+  // Parse the optional function body.
+  auto *body = result.addRegion();
+  if (parser.parseRegion(*body, entryArgs))
+    return failure();
+  if (body->empty())
+    body->push_back(new Block());
+
+  return success();
+}
+
+LogicalResult ClassOp::verify() {
+
+  for (auto operand : getBodyBlock()->getArguments()) {
+    auto type = operand.getType();
+    if (!isa<PropertyType>(type)) {
+      emitOpError("ports on a class must be properties");
+      return failure();
+    }
+  }
+
```

Comment:
Should this walk the body looking for "property ops" (however we define that)? I guess in the FIRRTL dialect, we are pretty rigorous about using HasParent in ODS, but currently something that was added without HasParent might slip in. It might be fine to have this as-is for now, since we don't have a nice trait to define "property ops".

Reply:
Yes I was thinking the same thing, "ParentOneOf" is idiomatic but not "good" enough. I think I will do that in a separate PR, where I will probably add some trait for propertyland ops. I think then, the HardwareDeclOp tg class (which I am adding in this PR) will be obsoleted.

Reply:
👍 that plan sounds good. I guess we could keep HardwareDeclOp?, add some kind of helper ODS class for PropertyOp that adds a trait, use it for PropAssignOp, and have the verifier here look for that trait.